syntax = "proto3";

package safe;

import "google/protobuf/timestamp.proto";

service Safe {
  rpc ListIssuers (ListIssuersRequest) returns (IssuerList);

  rpc CreateIssuer (CreateIssuerRequest) returns (CreateIssuerResponse);

  // The client secrets do not directly stem from the issuer's private key, but
  // form a symmetric key that in turn encrypts the private key. This allows us
  // to roll the client secrets.
  rpc RollClientSecrets (RollClientSecretsRequest) returns (RollClientSecretsResponse);

  rpc UpdateCrl (UpdateCrlRequest) returns (UpdateCrlResponse);

  rpc SignCertificate (SignCertificateRequest) returns (SignCertificateResponse);

  rpc RevokeCertificate (RevokeCertificateRequest) returns (RevokeCertificateResponse);

  rpc ListCertificates (ListCertificatesRequest) returns (ListCertificatesResponse);
}

message ListIssuersRequest {}

message IssuerList {
  repeated IssuerInfo issuers = 1;
}

message CreateIssuerRequest {
    string identifier = 1;
    string cert = 2;
    string private_key = 3;
    optional uint32 n_client_secrets = 4;
}

message CreateIssuerResponse {
    string identifier = 1;
    string cert = 2;
    repeated string client_secrets = 3;
}

message IssuerInfo {
  string identifier = 1;
  // PEM-encoded certificate
  string cert = 2;
}

message RollClientSecretsRequest {
  string issuer = 1;
  string secret = 2;
  optional uint32 n_client_secrets = 3;
}

message RollClientSecretsResponse {
  repeated string client_secrets = 1;
}

message UpdateCrlRequest {
  string issuer = 1;
  string secret = 2;
}

message UpdateCrlResponse {}

message SignCertificateRequest {
  string issuer = 1;
  string secret = 2;
  // DER-encoded certificate signing request.
  bytes csr = 3;
  optional google.protobuf.Timestamp not_before = 4;
  optional google.protobuf.Timestamp not_after = 5;
}

message SignCertificateResponse {
  bytes der = 1;
}

enum RevocationReason {
  UNSPECIFIED = 0;
  KEY_COMPROMISE = 1;
  CA_COMPROMISE = 2;
  AFFILIATION_CHANGED = 3;
  SUPERSEDED = 4;
  CESSATION_OF_OPERATION = 5;
  CERTIFICATE_HOLD = 6;
  REMOVE_FROM_CRL = 8;
  PRIVILEGE_WITHDRAWN = 9;
  AA_COMPROMISE = 10;
}

message RevokeCertificateRequest {
  string issuer = 1;
  string secret = 2;
  string serial = 3;
  optional RevocationReason reason = 4;
}

message RevokeCertificateResponse {}

message ListCertificatesRequest {
  string issuer = 1;
}

message ListCertificatesResponse {
  repeated string certificates = 1;
}
